java-checks/src/main/java/org/sonar/java/checks/FileCreateTempFileCheck.java

```diff
@@ -0,0 +1,149 @@
+  tags = {"owasp-a9", "security"},
+  priority = Priority.CRITICAL)
+@ActivatedByDefault
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.API_ABUSE)
+@SqaleConstantRemediation("5min")
+public class FileCreateTempFileCheck extends BaseTreeVisitor implements JavaFileScanner {
+
+  private enum State {
+    CREATE_TMP_FILE,
+    DELETE,
+    MKDIR
+  }
+
+  private static final String JAVA_IO_FILE = "java.io.File";
+  private static final MethodInvocationMatcher FILE_CREATE_TEMP_FILE = MethodInvocationMatcher.create()
+    .typeDefinition(JAVA_IO_FILE).name("createTempFile").withNoParameterConstraint();
+  private static final MethodInvocationMatcher FILE_DELETE = MethodInvocationMatcher.create()
+    .typeDefinition(JAVA_IO_FILE).name("delete");
+  private static final MethodInvocationMatcher FILE_MKDIR = MethodInvocationMatcher.create()
+    .typeDefinition(JAVA_IO_FILE).name("mkdir");
+
+  private final Deque<Map<Symbol, State>> symbolStack = new LinkedList<>();
+  private JavaFileScannerContext context;
+
+  @Override
+  public void scanFile(final JavaFileScannerContext context) {
+    this.context = context;
+    scan(context.getTree());
+  }
+
+  @Override
+  public void visitMethod(MethodTree tree) {
+    symbolStack.push(new HashMap<Symbol, State>());
+    super.visitMethod(tree);
+    symbolStack.pop();
+  }
+
+  @Override
+  public void visitAssignmentExpression(AssignmentExpressionTree tree) {
+    super.visitAssignmentExpression(tree);
+    if (isFileCreateTempFile(tree.expression())) {
+      ExpressionTree variable = tree.variable();
+      if (variable.is(Tree.Kind.IDENTIFIER) && !symbolStack.isEmpty()) {
+        symbolStack.peek().put(((IdentifierTree) variable).symbol(), State.CREATE_TMP_FILE);
+      }
+    }
+  }
+
+  @Override
+  public void visitVariable(VariableTree tree) {
+    super.visitVariable(tree);
+    ExpressionTree initializer = tree.initializer();
+    if (initializer != null && isFileCreateTempFile(initializer)) {
+      Symbol symbol = tree.symbol();
+      if (!symbolStack.isEmpty()) {
+        symbolStack.peek().put(symbol, State.CREATE_TMP_FILE);
+      }
+    }
+  }
+
+  private boolean isFileCreateTempFile(ExpressionTree givenExpression) {
+    ExpressionTree expressionTree = removeParenthesis(givenExpression);
+    return expressionTree.is(Tree.Kind.METHOD_INVOCATION) && FILE_CREATE_TEMP_FILE.matches((MethodInvocationTree) expressionTree);
+  }
+
+  @Override
+  public void visitMethodInvocation(MethodInvocationTree mit) {
+    super.visitMethodInvocation(mit);
+    if (FILE_DELETE.matches(mit)) {
+      checkState(mit, State.CREATE_TMP_FILE, State.DELETE);
+    } else if (FILE_MKDIR.matches(mit) && State.MKDIR.equals(checkState(mit, State.DELETE, State.MKDIR))) {
+      context.addIssue(mit, this, "Use \"Files.createTempDirectory\" or a library function to create this directory instead.");
+    }
+  }
+
+  private State checkState(MethodInvocationTree mit, State requiredState, State nextState) {
```

Comment:
this method can return null so it should be annotated at least with @Nullable or @CheckForNull